pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -302,6 +306,124 @@ func (m *kubeGenericRuntimeManager) getKubeletContainerStatuses(podSandboxID str
 	return statuses, nil
 }
 
+// generateContainerEvent generates an event for the container.
+func (m *kubeGenericRuntimeManager) generateContainerEvent(containerID kubecontainer.ContainerID, eventType, reason, message string) {
+	ref, ok := m.containerRefManager.GetRef(containerID)
+	if !ok {
+		glog.Warningf("No ref for container %q", containerID)
+		return
+	}
+	m.recorder.Event(ref, eventType, reason, message)
+}
+
+// executePreStopHook runs the pre-stop lifecycle hooks if applicable and returns new gracePeriod.
+func (m *kubeGenericRuntimeManager) executePreStopHook(pod *api.Pod, containerID kubecontainer.ContainerID, containerSpec *api.Container, gracePeriod int64) int64 {
+	start := unversioned.Now()
+	if pod != nil && containerSpec != nil && containerSpec.Lifecycle != nil && containerSpec.Lifecycle.PreStop != nil {
```

Comment:
Couple nits about how this was broken out:
I would rather this check not be in this function, but in the caller.
It saves us a level of indentation, and it lets us make this function have a tighter responsibility.

I also think it's a little easier to follow if the function returns the duration the hook takes (and the grace period subtraction is done a level up) and whether the hook succeeded or failed (and the `genereateContainerEvent` is done by the caller as well)